CI: fsspec 2021.6.0 failures

`pandas/tests/io/test_fsspec.py::test_read_csv` and `pandas/tests/io/test_fsspec.py::test_markdown_options` fails with fsspec version 2021.6.0.

Based on the changelog, the culprit might be 
> Better testing and folder handling for Memory (654)

but more investigation needed to see if this is a new bug or something we should fix in the failing tests

Temporary fix in #42023

For reference, the failures (indeed related to MemoryFileSystem):

```
=================================== FAILURES ===================================
________________________________ test_read_csv _________________________________
[gw1] linux -- Python 3.7.10 /usr/share/miniconda/envs/pandas-dev/bin/python

cleared_fs = <fsspec.implementations.memory.MemoryFileSystem object at 0x7f7e66ebff90>

    def test_read_csv(cleared_fs):
        from fsspec.implementations.memory import MemoryFile
    
        cleared_fs.store["test/test.csv"] = MemoryFile(data=text)
>       df2 = read_csv("memory://test/test.csv", parse_dates=["dt"])

pandas/tests/io/test_fsspec.py:44: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
pandas/util/_decorators.py:311: in wrapper
    return func(*args, **kwargs)
pandas/io/parsers/readers.py:586: in read_csv
    return _read(filepath_or_buffer, kwds)
pandas/io/parsers/readers.py:482: in _read
    parser = TextFileReader(filepath_or_buffer, **kwds)
pandas/io/parsers/readers.py:811: in __init__
    self._engine = self._make_engine(self.engine)
pandas/io/parsers/readers.py:1040: in _make_engine
    return mapping[engine](self.f, **self.options)  # type: ignore[call-arg]
pandas/io/parsers/c_parser_wrapper.py:51: in __init__
    self._open_handles(src, kwds)
pandas/io/parsers/base_parser.py:228: in _open_handles
    errors=kwds.get("encoding_errors", "strict"),
pandas/io/common.py:613: in get_handle
    storage_options=storage_options,
pandas/io/common.py:358: in _get_filepath_or_buffer
    filepath_or_buffer, mode=fsspec_mode, **(storage_options or {})
/usr/share/miniconda/envs/pandas-dev/lib/python3.7/site-packages/fsspec/core.py:135: in open
    out = self.__enter__()
/usr/share/miniconda/envs/pandas-dev/lib/python3.7/site-packages/fsspec/core.py:102: in __enter__
    f = self.fs.open(self.path, mode=mode)
/usr/share/miniconda/envs/pandas-dev/lib/python3.7/site-packages/fsspec/spec.py:968: in open
    **kwargs,
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <fsspec.implementations.memory.MemoryFileSystem object at 0x7f7e66ebff90>
path = '/test/test.csv', mode = 'rb', block_size = None, autocommit = True
cache_options = None, kwargs = {}, parent = '/'

    def _open(
        self,
        path,
        mode="rb",
        block_size=None,
        autocommit=True,
        cache_options=None,
        **kwargs,
    ):
        path = self._strip_protocol(path)
        if path in self.pseudo_dirs:
            raise IsADirectoryError
        parent = path
        while len(parent) > 1:
            parent = self._parent(parent)
            if self.isfile(parent):
                raise FileExistsError(parent)
        if mode in ["rb", "ab", "rb+"]:
            if path in self.store:
                f = self.store[path]
                if mode == "ab":
                    # position at the end of file
                    f.seek(0, 2)
                else:
                    # position at the beginning of file
                    f.seek(0)
                return f
            else:
>               raise FileNotFoundError(path)
E               FileNotFoundError: /test/test.csv

/usr/share/miniconda/envs/pandas-dev/lib/python3.7/site-packages/fsspec/implementations/memory.py:179: FileNotFoundError
____________________________ test_markdown_options _____________________________
[gw1] linux -- Python 3.7.10 /usr/share/miniconda/envs/pandas-dev/bin/python

self = <pandas.conftest.fsspectest.<locals>.TestMemoryFS object at 0x7f7e62034110>
path = '/afile', start = None, end = None, kwargs = {}

    def cat_file(self, path, start=None, end=None, **kwargs):
        path = self._strip_protocol(path)
        try:
>           return self.store[path].getvalue()[start:end]
E           KeyError: '/afile'

/usr/share/miniconda/envs/pandas-dev/lib/python3.7/site-packages/fsspec/implementations/memory.py:200: KeyError

During handling of the above exception, another exception occurred:

fsspectest = <pandas.conftest.fsspectest.<locals>.TestMemoryFS object at 0x7f7e62034110>

    @td.skip_if_no("tabulate")
    def test_markdown_options(fsspectest):
        df = DataFrame({"a": [0]})
        df.to_markdown("testmem://afile", storage_options={"test": "md_write"})
        assert fsspectest.test[0] == "md_write"
>       assert fsspectest.cat("afile")

pandas/tests/io/test_fsspec.py:285: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
/usr/share/miniconda/envs/pandas-dev/lib/python3.7/site-packages/fsspec/spec.py:727: in cat
    return self.cat_file(paths[0], **kwargs)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <pandas.conftest.fsspectest.<locals>.TestMemoryFS object at 0x7f7e62034110>
path = '/afile', start = None, end = None, kwargs = {}

    def cat_file(self, path, start=None, end=None, **kwargs):
        path = self._strip_protocol(path)
        try:
            return self.store[path].getvalue()[start:end]
        except KeyError:
>           raise FileNotFoundError(path)
E           FileNotFoundError: /afile

/usr/share/miniconda/envs/pandas-dev/lib/python3.7/site-packages/fsspec/implementations/memory.py:202: FileNotFoundError
```

--- a/ci/deps/actions-38-db.yaml
+++ b/ci/deps/actions-38-db.yaml
@@ -16,7 +16,7 @@ dependencies:
   - botocore>=1.11
   - dask
   - fastparquet>=0.4.0
-  - fsspec>=0.7.4, <2021.6.0
+  - fsspec>=0.7.4
   - gcsfs>=0.6.0
   - geopandas
   - html5lib

--- a/ci/deps/actions-38-slow.yaml
+++ b/ci/deps/actions-38-slow.yaml
@@ -13,7 +13,7 @@ dependencies:
 
   # pandas dependencies
   - beautifulsoup4
-  - fsspec>=0.7.4, <2021.6.0
+  - fsspec>=0.7.4
   - html5lib
   - lxml
   - matplotlib

--- a/ci/deps/actions-39-slow.yaml
+++ b/ci/deps/actions-39-slow.yaml
@@ -15,7 +15,7 @@ dependencies:
   # pandas dependencies
   - beautifulsoup4
   - bottleneck
-  - fsspec>=0.8.0, <2021.6.0
+  - fsspec>=0.8.0
   - gcsfs
   - html5lib
   - jinja2

--- a/ci/deps/actions-39.yaml
+++ b/ci/deps/actions-39.yaml
@@ -14,7 +14,7 @@ dependencies:
   # pandas dependencies
   - beautifulsoup4
   - bottleneck
-  - fsspec>=0.8.0, <2021.6.0
+  - fsspec>=0.8.0
   - gcsfs
   - html5lib
   - jinja2

--- a/ci/deps/azure-windows-38.yaml
+++ b/ci/deps/azure-windows-38.yaml
@@ -17,7 +17,7 @@ dependencies:
   - bottleneck
   - fastparquet>=0.4.0
   - flask
-  - fsspec>=0.8.0, <2021.6.0
+  - fsspec>=0.8.0
   - matplotlib=3.3.2
   - moto>=1.3.14
   - numba

--- a/ci/deps/azure-windows-39.yaml
+++ b/ci/deps/azure-windows-39.yaml
@@ -15,7 +15,7 @@ dependencies:
   # pandas dependencies
   - beautifulsoup4
   - bottleneck
-  - fsspec>=0.8.0, <2021.6.0
+  - fsspec>=0.8.0
   - gcsfs
   - html5lib
   - jinja2

--- a/environment.yml
+++ b/environment.yml
@@ -106,7 +106,7 @@ dependencies:
   - pytables>=3.6.1  # pandas.read_hdf, DataFrame.to_hdf
   - s3fs>=0.4.0  # file IO when using 's3://...' path
   - aiobotocore
-  - fsspec>=0.7.4, <2021.6.0  # for generic remote file operations
+  - fsspec>=0.7.4  # for generic remote file operations
   - gcsfs>=0.6.0  # file IO when using 'gcs://...' path
   - sqlalchemy  # pandas.read_sql, DataFrame.to_sql
   - xarray<0.19  # DataFrame.to_xarray

--- a/pandas/tests/io/test_fsspec.py
+++ b/pandas/tests/io/test_fsspec.py
@@ -41,7 +41,7 @@ def cleared_fs():
 def test_read_csv(cleared_fs):
     from fsspec.implementations.memory import MemoryFile
 
-    cleared_fs.store["test/test.csv"] = MemoryFile(data=text)
+    cleared_fs.store["/test/test.csv"] = MemoryFile(data=text)
     df2 = read_csv("memory://test/test.csv", parse_dates=["dt"])
 
     tm.assert_frame_equal(df1, df2)

--- a/pandas/tests/io/test_fsspec.py
+++ b/pandas/tests/io/test_fsspec.py
@@ -38,12 +38,9 @@ def cleared_fs():
     memfs.store.clear()
 
 
-# Different behavior for before 2021.6.0
-@td.skip_if_no("fsspec", "2021.6.0")
 def test_read_csv(cleared_fs):
-    from fsspec.implementations.memory import MemoryFile
-
-    cleared_fs.store["/test/test.csv"] = MemoryFile(data=text)
+    with cleared_fs.open("test/test.csv", "wb") as w:
+        w.write(text)
     df2 = read_csv("memory://test/test.csv", parse_dates=["dt"])
 
     tm.assert_frame_equal(df1, df2)

--- a/pandas/tests/io/test_fsspec.py
+++ b/pandas/tests/io/test_fsspec.py
@@ -38,6 +38,8 @@ def cleared_fs():
     memfs.store.clear()
 
 
+# Different behavior for before 2021.6.0
+@td.skip_if_no("fsspec", "2021.6.0")
 def test_read_csv(cleared_fs):
     from fsspec.implementations.memory import MemoryFile
 
@@ -294,7 +296,7 @@ def test_markdown_options(fsspectest):
     df = DataFrame({"a": [0]})
     df.to_markdown("testmem://afile", storage_options={"test": "md_write"})
     assert fsspectest.test[0] == "md_write"
-    assert fsspectest.cat("afile")
+    assert fsspectest.cat("testmem://afile")
 
 
 @td.skip_if_no("pyarrow")

--- a/requirements-dev.txt
+++ b/requirements-dev.txt
@@ -70,7 +70,7 @@ python-snappy
 tables>=3.6.1
 s3fs>=0.4.0
 aiobotocore
-fsspec>=0.7.4, <2021.6.0
+fsspec>=0.7.4
 gcsfs>=0.6.0
 sqlalchemy
 xarray<0.19

